### PR TITLE
fix(core): Perform multi-main leader check against key ID

### PR DESCRIPTION
**Pull request by ivov**
## Summary
Current leader check is based on key presence and multi-main instance type, which can give rise to [this edge case](https://n8nio.slack.com/archives/C04B1GZ4T0U/p1702025497086379?thread_ts=1701962808.817579&cid=C04B1GZ4T0U) where leader fails to realize they lost leadership to a former follower. PR performs the check against the specific key ID instead to prevent this.

...

#### How to test the change:
1. ...

## Issues fixed
Include links to Github issue or Community forum post or **Linear ticket**:
> Important in order to close automatically and provide context to reviewers

...

## Review / Merge checklist
- [ ] PR title and summary are descriptive. **Remember, the title automatically goes into the changelog. Use `(no-changelog)` otherwise.** ([conventions](https://github.com/n8n-io/n8n/blob/master/.github/pull_request_title_conventions.md))
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [ ] Tests included.
   > A bug is not considered fixed, unless a test is added to prevent it from happening again. A feature is not complete without tests. 
  >
  > *(internal)* You can use Slack commands to trigger [e2e tests](https://www.notion.so/n8n/How-to-use-Test-Instances-d65f49dfc51f441ea44367fb6f67eb0a?pvs=4#a39f9e5ba64a48b58a71d81c837e8227) or [deploy test instance](https://www.notion.so/n8n/How-to-use-Test-Instances-d65f49dfc51f441ea44367fb6f67eb0a?pvs=4#f6a177d32bde4b57ae2da0b8e454bfce) or [deploy early access version on Cloud](https://www.notion.so/n8n/Cloudbot-3dbe779836004972b7057bc989526998?pvs=4#fef2d36ab02247e1a0f65a74f6fb534e).

### CODE REVIEW

File: packages/cli/src/services/orchestration/main/MultiMainSetup.ee.ts [hunk 1]
```diff
@@ -63,41 +63,43 @@ export class MultiMainSetup extends SingleMainSetup {
...
-
 		const leaderId = await this.redisPublisher.get(this.leaderKey);
 
-		if (!leaderId) {
-			this.logger.debug('Leadership vacant, attempting to become leader...');
-			await this.tryBecomeLeader();
+		if (leaderId === this.id) {
+			this.logger.debug(`[Instance ID ${this.id}] Leader is this instance`);
+
+			config.set('multiMainSetup.instanceType', 'leader');
```
Comment (krynble) [inline packages/cli/src/services/orchestration/main/MultiMainSetup.ee.ts:71]: Maybe we need to check if not already the leader, we should fire the event stating that this instance became the leader; isn't this necessary?

Have you tried killing the current leader to see what happens? Was the transition ok?

Comment (ivov) [reply packages/cli/src/services/orchestration/main/MultiMainSetup.ee.ts:71]: > Maybe we need to check if not already the leader, we should fire the event stating that this instance became the leader; isn't this necessary?

`leaderId === this.id` confirms that the instance _remained_ the leader, not that it _became_ the leader, so here there was no leader transition, so we do not emit a leadership change event. Unless you mean something else?

> Have you tried killing the current leader to see what happens? Was the transition ok?

I [manually tested](https://share.cleanshot.com/LLX90BFD) transitions. What is harder to test locally is spontaneous transitions.

Comment (krynble) [reply packages/cli/src/services/orchestration/main/MultiMainSetup.ee.ts:71]: I reviewed the code and it seems ok but this call to `config.set('multiMainSetup.instanceType', 'leader');` seems redundant.

File: packages/cli/src/services/orchestration/main/MultiMainSetup.ee.ts [hunk 2]
```diff
@@ -63,41 +63,43 @@ export class MultiMainSetup extends SingleMainSetup {
...
+
+			await this.redisPublisher.setExpiration(this.leaderKey, this.leaderKeyTtl);
 
 			return;
 		}
 
-		if (this.isLeader) {
-			this.logger.debug(`Leader is this instance "${this.id}"`);
+		if (leaderId && leaderId !== this.id) {
```
Comment (krynble) [inline packages/cli/src/services/orchestration/main/MultiMainSetup.ee.ts:78]: Even more importantly here, where the current instance realizes it is not the leader it's not firing `this.emit('leadershipChange', this.id);`

This means that even if it fails to maintain the lock and another leader takes over, the triggers here will never be stopped.

Comment (ivov) [reply packages/cli/src/services/orchestration/main/MultiMainSetup.ee.ts:78]: I may be missing something - but from what I understand `this.emit('leadershipChange', this.id);` is already fired by the follower who gained leadership.

- `leaderId === this.id` → leader remained leader, so mark as leader (redundant sanity check)
- `leaderId && leaderId !== this.id` → follower remained follower, so mark as follower (redundant sanity check)
- `!leaderId` → no leader, so attempt to set key
	- if successfully set key, mark as leader and extend lease and emit leadership change event
	- else, mark as follower

Maybe faster to discuss?

Comment (krynble) [reply packages/cli/src/services/orchestration/main/MultiMainSetup.ee.ts:78]: I think the path for becoming a leader is somewhat ok, the problem is when the instance loses leadership.

All we do is set `config.set('multiMainSetup.instanceType', 'follower');` but we're not disabling the triggers.

Makes sense?

File: packages/nodes-base/nodes/Transform/utils/utils.ts [hunk 3]
```diff
@@ -8,7 +10,8 @@ export const prepareFieldsArray = (fields: string | string[], fieldName = 'Field
 	if (Array.isArray(fields)) {
 		return fields;
 	}
-	throw new Error(
+	throw new ApplicationError(
```
Comment (krynble) [inline packages/nodes-base/nodes/Transform/utils/utils.ts:13]: I think this change does not belong here right?

Comment (ivov) [reply packages/nodes-base/nodes/Transform/utils/utils.ts:13]: Addressing here: https://github.com/n8n-io/n8n/pull/7966

**Review by krynble — APPROVED** [review]

**Comment by janober** [comment]
Got released with [`n8n@1.21.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.21.0)

**Pull request merged**